Added validation for collection filters

allouis opened the pull request.

allouis committed:
Validated filters for collections

refs https://github.com/TryGhost/Arch/issues/47

This ensures that we only have collections which have a valid filter in terms of
  - Valid NQL string
  - Uses only properties which are valid to filter on
  - Only has an empty filter in the case of the "latest" collection

allouis committed:
Removed invalid entities from results

refs https://github.com/TryGhost/Arch/issues/47

This change allows us to throw errors when instantiating invalid entities,
whilst not breaking things when we have bad data in the database. What we can
do is act as if the bad rows are not present, whilst surfacing an error in
sentry to alert us to such cases.

naz commented on ghost/core/core/server/services/collections/BookshelfCollectionsRepository.js:
```diff
@@ -61,24 +62,29 @@ module.exports = class BookshelfCollectionsRepository {
+                updatedAt: json.updated_at
+            });
+        } catch (err) {
+            sentry.captureException(err);
```
My understanding is that we want to surface these errors more prominently instead of throwing everything into one bag with `logger.error`. If that's the intention maybe we should create a different abstraction instead of depending on very concrete 3rd party like sentry?

allouis replied:
`sentry.captureException` doesn't log out the event, it captures it in sentry and we can setup alerts for it! 

RE: a different abstraction - I think potentially yes, but RN I think that's premature, we currently refer to sentry by name everywhere else, and given this is part of the application codebase (i.e not business logic) we're okay to continue

I think this is missing a `logger.error` though - so I'll add that

allouis committed:
Handled our forced null conversion in bookshelf

refs https://github.com/TryGhost/Arch/issues/47

We've configured bookshelf to force empty strings to null, but this is
undesired behaviour here, so unfortunately we have to leak some business
logic into the repository.

This needs to be done to correctly support our filter validation logic.
